fix: Missing replication_specs on mongodbatlas_advanced_cluster(s) data sources

drcapulet opened the pull request:
I couldn't quite understand what triggered the `Set: replicationSpecsHashSet` line in the resource to be commented out but making the data source match fixes the issue. Possibly because the `region_configs` set would also need a custom `Set` function?, but either way there aren't any tests that caught this and I don't have an Atlas organization I can run the tests against to add one. I did install the provider locally and ran a plan to confirm this fixes the issue in my Terraformm .One likely reason is that none of the examples seem to have similar configs across multiple regions (like two with two electable nodes each).

## Description

Please include a summary of the fix/feature/change, including any relevant motivation and context.

Link to any related issue(s):

## Type of change:

- [x] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

## Further comments

marcosuma commented:
Hi @drcapulet, thanks so much for opening a PR with us - just to confirm: you've identified a bug in the advanced_cluster data source for which a replication_specs doesn't get returned in the results. Is that correct?

If so, would you be so kind to give us the example of your structure so we could use it to replicate the issue and add tests around it? Of course, please remove any non-necessary sensitive information - we just need your cluster structure that once fetched through data source causes the bug.

drcapulet commented:
@marcosuma Indeed - that's correct! Here's a resource definition that should trigger it:

```terraform
resource "mongodbatlas_advanced_cluster" "example_data_failure" {
  cluster_type = "REPLICASET"
  name         = "example-data-failure"
  project_id   = mongodbatlas_project.main.id

  replication_specs {
    num_shards = 1

    region_configs {
      provider_name = "GCP"
      priority      = 7
      region_name   = "NORTH_AMERICA_NORTHEAST_1"

      auto_scaling {
        compute_enabled            = false
        compute_scale_down_enabled = false
        disk_gb_enabled            = true
      }

      electable_specs {
        instance_size = "M10"
        node_count    = 3
      }
    }

    dynamic "region_configs" {
      for_each = [
        "EASTERN_US",
        "US_SOUTH_1",
        "US_WEST_2",
      ]

      content {
        provider_name = "GCP"
        priority      = 0
        region_name   = region_configs.value

        auto_scaling {
          compute_enabled            = false
          compute_scale_down_enabled = false
          disk_gb_enabled            = true
        }

        read_only_specs {
          instance_size = "M10"
          node_count    = 1
        }
      }
    }
  }
}
```

marcosuma commented:
I think I managed to replicate this issue with the following example:

```
resource "mongodbatlas_advanced_cluster" "example_data_failure" {
  cluster_type = "REPLICASET"
  name         = "example-data-failure-2"
  project_id   = var.project_id

  replication_specs {
    num_shards = 1

    region_configs {
      provider_name = "GCP"
      priority      = 7
      region_name   = "US_WEST_2"

      auto_scaling {
        compute_enabled            = false
        compute_scale_down_enabled = false
        disk_gb_enabled            = true
      }

      electable_specs {
        instance_size = "M10"
        node_count    = 3
      }
    }

    dynamic "region_configs" {
      for_each = [
        # "US_EAST_4",
        # "US_SOUTH_1",
        "US_EAST_4",
        "US_WEST_3"
      ]

      content {
        provider_name = "GCP"
        priority      = 0
        region_name   = region_configs.value

        auto_scaling {
          compute_enabled            = false
          compute_scale_down_enabled = false
          disk_gb_enabled            = true
        }

        read_only_specs {
          instance_size = "M10"
          node_count    = 1
        }
      }
    }
  }
}

data "mongodbatlas_advanced_cluster" "example_data_failure" {
  name         = "example-data-failure-2"
  project_id   = var.project_id
}

output "mongodbatlas_advanced_cluster_replication_specs" {
  value = data.mongodbatlas_advanced_cluster.example_data_failure.replication_specs
}
```

and I get a response like:

```
data.mongodbatlas_advanced_cluster.example_data_failure: Reading...
mongodbatlas_advanced_cluster.example_data_failure: Refreshing state... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NGE1YzhjZmQw-Y2x1c3Rlcl9uYW1l:ZXhhbXBsZS1kYXRhLWZhaWx1cmUtMg==-cHJvamVjdF9pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz]
data.mongodbatlas_advanced_cluster.example_data_failure: Read complete after 2s [id=6618e67f11bcd974a5c8cfd0]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # mongodbatlas_advanced_cluster.example_data_failure will be updated in-place
  ~ resource "mongodbatlas_advanced_cluster" "example_data_failure" {
        id                             = "Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NGE1YzhjZmQw-Y2x1c3Rlcl9uYW1l:ZXhhbXBsZS1kYXRhLWZhaWx1cmUtMg==-cHJvamVjdF9pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz"
        name                           = "example-data-failure-2"
        # (16 unchanged attributes hidden)

      ~ replication_specs {
            id           = "6618e67f11bcd974a5c8cfc8"
            # (3 unchanged attributes hidden)

          ~ region_configs {
              ~ region_name   = "US_WEST_3" -> "US_EAST_4"
                # (2 unchanged attributes hidden)

                # (2 unchanged blocks hidden)
            }
          ~ region_configs {
              ~ region_name   = "US_EAST_4" -> "US_WEST_3"
                # (2 unchanged attributes hidden)

                # (2 unchanged blocks hidden)
            }

            # (1 unchanged block hidden)
        }

        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Changes to Outputs:
  + mongodbatlas_advanced_cluster_replication_specs = [
      + {
          + container_id   = {
              + "GCP:US_EAST_4" = "6618e36d11bcd974a5c8cc8e"
              + "GCP:US_WEST_2" = "6618e36d11bcd974a5c8cc8e"
              + "GCP:US_WEST_3" = "6618e36d11bcd974a5c8cc8e"
            }
          + id             = "6618e67f11bcd974a5c8cfc8"
          + num_shards     = 1
          + region_configs = [
              + {
                  + analytics_auto_scaling = []
                  + analytics_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + auto_scaling           = [
                      + {
                          + compute_enabled            = false
                          + compute_max_instance_size  = ""
                          + compute_min_instance_size  = ""
                          + compute_scale_down_enabled = false
                          + disk_gb_enabled            = true
                        },
                    ]
                  + backing_provider_name  = ""
                  + electable_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + priority               = 0
                  + provider_name          = "GCP"
                  + read_only_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 1
                        },
                    ]
                  + region_name            = "US_WEST_3"
                },
              + {
                  + analytics_auto_scaling = []
                  + analytics_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + auto_scaling           = [
                      + {
                          + compute_enabled            = false
                          + compute_max_instance_size  = ""
                          + compute_min_instance_size  = ""
                          + compute_scale_down_enabled = false
                          + disk_gb_enabled            = true
                        },
                    ]
                  + backing_provider_name  = ""
                  + electable_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 3
                        },
                    ]
                  + priority               = 7
                  + provider_name          = "GCP"
                  + read_only_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + region_name            = "US_WEST_2"
                },
            ]
          + zone_name      = "ZoneName managed by Terraform"
        },
    ]

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

mongodbatlas_advanced_cluster.example_data_failure: Modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NGE1YzhjZmQw-Y2x1c3Rlcl9uYW1l:ZXhhbXBsZS1kYXRhLWZhaWx1cmUtMg==-cHJvamVjdF9pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 10s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 20s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 30s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 40s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 50s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Still modifying... [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NG...pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz, 1m0s elapsed]
mongodbatlas_advanced_cluster.example_data_failure: Modifications complete after 1m3s [id=Y2x1c3Rlcl9pZA==:NjYxOGU2N2YxMWJjZDk3NGE1YzhjZmQw-Y2x1c3Rlcl9uYW1l:ZXhhbXBsZS1kYXRhLWZhaWx1cmUtMg==-cHJvamVjdF9pZA==:NjVmODJjNGRkNGM5MWU2YzYxNDM3NjIz]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

Outputs:

mongodbatlas_advanced_cluster_replication_specs = toset([
  {
    "container_id" = tomap({
      "GCP:US_EAST_4" = "6618e36d11bcd974a5c8cc8e"
      "GCP:US_WEST_2" = "6618e36d11bcd974a5c8cc8e"
      "GCP:US_WEST_3" = "6618e36d11bcd974a5c8cc8e"
    })
    "id" = "6618e67f11bcd974a5c8cfc8"
    "num_shards" = 1
    "region_configs" = toset([
      {
        "analytics_auto_scaling" = tolist([])
        "analytics_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "auto_scaling" = tolist([
          {
            "compute_enabled" = false
            "compute_max_instance_size" = ""
            "compute_min_instance_size" = ""
            "compute_scale_down_enabled" = false
            "disk_gb_enabled" = true
          },
        ])
        "backing_provider_name" = ""
        "electable_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "priority" = 0
        "provider_name" = "GCP"
        "read_only_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 1
          },
        ])
        "region_name" = "US_WEST_3"
      },
      {
        "analytics_auto_scaling" = tolist([])
        "analytics_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "auto_scaling" = tolist([
          {
            "compute_enabled" = false
            "compute_max_instance_size" = ""
            "compute_min_instance_size" = ""
            "compute_scale_down_enabled" = false
            "disk_gb_enabled" = true
          },
        ])
        "backing_provider_name" = ""
        "electable_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 3
          },
        ])
        "priority" = 7
        "provider_name" = "GCP"
        "read_only_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "region_name" = "US_WEST_2"
      },
    ])
    "zone_name" = "ZoneName managed by Terraform"
  },
])
```

Two issues:
1. every time I run a terraform plan, a plan change is detected between the two region names that are in the dynamic block. @drcapulet did this happen to you?
2. the output of the replication spec only contains two elems and not one, which I believe is the bug mentioned in this PR. @drcapulet can you confirm?

marcosuma commented:
The point 1. is addressed by inverting the order in the `for_each` - I believe this is an expected behaviour due to the `ListType`

marcosuma commented:
I can see how with these changes in place I can get the right output in full:

```
Changes to Outputs:
  ~ mongodbatlas_advanced_cluster_replication_specs = [
      ~ {
            id             = "6618e67f11bcd974a5c8cfc8"
          ~ region_configs = [
              - {
                  - analytics_auto_scaling = []
                  - analytics_specs        = [
                      - {
                          - disk_iops       = 0
                          - ebs_volume_type = ""
                          - instance_size   = "M10"
                          - node_count      = 0
                        },
                    ]
                  - auto_scaling           = [
                      - {
                          - compute_enabled            = false
                          - compute_max_instance_size  = ""
                          - compute_min_instance_size  = ""
                          - compute_scale_down_enabled = false
                          - disk_gb_enabled            = true
                        },
                    ]
                  - backing_provider_name  = ""
                  - electable_specs        = [
                      - {
                          - disk_iops       = 0
                          - ebs_volume_type = ""
                          - instance_size   = "M10"
                          - node_count      = 0
                        },
                    ]
                  - priority               = 0
                  - provider_name          = "GCP"
                  - read_only_specs        = [
                      - {
                          - disk_iops       = 0
                          - ebs_volume_type = ""
                          - instance_size   = "M10"
                          - node_count      = 1
                        },
                    ]
                  - region_name            = "US_WEST_3"
                },
                {
                    analytics_auto_scaling = []
                    analytics_specs        = [
                        {
                            disk_iops       = 0
                            ebs_volume_type = ""
                            instance_size   = "M10"
                            node_count      = 0
                        },
                    ]
                    auto_scaling           = [
                        {
                            compute_enabled            = false
                            compute_max_instance_size  = ""
                            compute_min_instance_size  = ""
                            compute_scale_down_enabled = false
                            disk_gb_enabled            = true
                        },
                    ]
                    backing_provider_name  = ""
                    electable_specs        = [
                        {
                            disk_iops       = 0
                            ebs_volume_type = ""
                            instance_size   = "M10"
                            node_count      = 3
                        },
                    ]
                    priority               = 7
                    provider_name          = "GCP"
                    read_only_specs        = [
                        {
                            disk_iops       = 0
                            ebs_volume_type = ""
                            instance_size   = "M10"
                            node_count      = 0
                        },
                    ]
                    region_name            = "US_WEST_2"
                },
              + {
                  + analytics_auto_scaling = []
                  + analytics_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + auto_scaling           = [
                      + {
                          + compute_enabled            = false
                          + compute_max_instance_size  = ""
                          + compute_min_instance_size  = ""
                          + compute_scale_down_enabled = false
                          + disk_gb_enabled            = true
                        },
                    ]
                  + backing_provider_name  = ""
                  + electable_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + priority               = 0
                  + provider_name          = "GCP"
                  + read_only_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 1
                        },
                    ]
                  + region_name            = "US_WEST_3"
                },
              + {
                  + analytics_auto_scaling = []
                  + analytics_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + auto_scaling           = [
                      + {
                          + compute_enabled            = false
                          + compute_max_instance_size  = ""
                          + compute_min_instance_size  = ""
                          + compute_scale_down_enabled = false
                          + disk_gb_enabled            = true
                        },
                    ]
                  + backing_provider_name  = ""
                  + electable_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 0
                        },
                    ]
                  + priority               = 0
                  + provider_name          = "GCP"
                  + read_only_specs        = [
                      + {
                          + disk_iops       = 0
                          + ebs_volume_type = ""
                          + instance_size   = "M10"
                          + node_count      = 1
                        },
                    ]
                  + region_name            = "US_EAST_4"
                },
            ]
            # (3 unchanged attributes hidden)
        },
    ]

```

marcosuma commented:
```
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

mongodbatlas_advanced_cluster_replication_specs = tolist([
  {
    "container_id" = tomap({
      "GCP:US_EAST_4" = "6618e36d11bcd974a5c8cc8e"
      "GCP:US_WEST_2" = "6618e36d11bcd974a5c8cc8e"
      "GCP:US_WEST_3" = "6618e36d11bcd974a5c8cc8e"
    })
    "id" = "6618e67f11bcd974a5c8cfc8"
    "num_shards" = 1
    "region_configs" = tolist([
      {
        "analytics_auto_scaling" = tolist([])
        "analytics_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "auto_scaling" = tolist([
          {
            "compute_enabled" = false
            "compute_max_instance_size" = ""
            "compute_min_instance_size" = ""
            "compute_scale_down_enabled" = false
            "disk_gb_enabled" = true
          },
        ])
        "backing_provider_name" = ""
        "electable_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 3
          },
        ])
        "priority" = 7
        "provider_name" = "GCP"
        "read_only_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "region_name" = "US_WEST_2"
      },
      {
        "analytics_auto_scaling" = tolist([])
        "analytics_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "auto_scaling" = tolist([
          {
            "compute_enabled" = false
            "compute_max_instance_size" = ""
            "compute_min_instance_size" = ""
            "compute_scale_down_enabled" = false
            "disk_gb_enabled" = true
          },
        ])
        "backing_provider_name" = ""
        "electable_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "priority" = 0
        "provider_name" = "GCP"
        "read_only_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 1
          },
        ])
        "region_name" = "US_WEST_3"
      },
      {
        "analytics_auto_scaling" = tolist([])
        "analytics_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "auto_scaling" = tolist([
          {
            "compute_enabled" = false
            "compute_max_instance_size" = ""
            "compute_min_instance_size" = ""
            "compute_scale_down_enabled" = false
            "disk_gb_enabled" = true
          },
        ])
        "backing_provider_name" = ""
        "electable_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 0
          },
        ])
        "priority" = 0
        "provider_name" = "GCP"
        "read_only_specs" = tolist([
          {
            "disk_iops" = 0
            "ebs_volume_type" = ""
            "instance_size" = "M10"
            "node_count" = 1
          },
        ])
        "region_name" = "US_EAST_4"
      },
    ])
    "zone_name" = "ZoneName managed by Terraform"
  },
])
```

marcosuma commented:
@drcapulet you have a failing task because you have to run `make check-changelog-entry-file` ( you need to rebase/merge the latest changes from master to execute that command)

drcapulet commented:
@marcosuma Correct on both earlier points - I had fixed that in the TF we had but missed it when I simplified it down to the example I provided - updated with the changelog entry!

marcosuma commented on .changelog/2076.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:bug
+data-source/mongodbatlas_advanced_cluster: Fixes missing replication_specs
````
```suggestion
data-source/mongodbatlas_advanced_cluster: Converts replication_specs to TypeList
```

AgustinBettati replied:
Would additionally include in backticks for referring to the attribute: 
```data-source/mongodbatlas_advanced_cluster: Converts `replication_specs` to TypeList```

marcosuma commented on internal/service/advancedcluster/data_source_advanced_cluster.go:
```diff
@@ -208,7 +208,7 @@ func DataSource() *schema.Resource {
 					},
 				},
-				Set: replicationSpecsHashSet,
+				// Set: replicationSpecsHashSet,
```
let's just remove this (I know in the resource it is commented, but really don't know why 😃)

marcosuma commented:
@drcapulet I worked on what the additional test to write would be in here: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2145/files

feel free to copy/paste into this PR so you can still remain as the author of it. (note that we also have to update the plural data_source)

drcapulet committed:
fix: Missing replication_specs on mongodbatlas_advanced_cluster(s) data sources

I couldn't quite understand what triggered the `Set: replicationSpecsHashSet`
line in the resource to be commented out but making the data source match
fixes the issue. Possibly because the `region_configs` set would also need
a custom `Set` function?, but either way there aren't any tests that caught
this and I don't have an Atlas organization I can run the tests against to
add one. I did install the provider locally and ran a plan to confirm this
fixes the issue in my Terraformm .One likely reason is that none of the
examples seem to have similar configs across multiple regions (like two
with two electable nodes each).

drcapulet commented:
@marcosuma Appreciate it! Added that in + also added a test on the singular data source too.

marcosuma commented:
Awesome - running tests now. Once succeed I will approve, I will ask someone else in the team to also have a look since I contributed to this. Thanks so much @drcapulet, great collaboration.

marcosuma commented:
@drcapulet apologise for the back and forth - realised with the team that we don't support running our whole test infrastructure on forks for security reasons. That's why I've created https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2145 and cherry-picked your commit. You still remain the co-author of it, so your contribution doesn't get lost, which was my top of mind.

I am going to close this, feel free to jump in #2145 which will be soon reviewed by the team.

drcapulet commented:
@marcosuma No worries, I completely understand! Appreciate the attention to the co-authorship & I'll watch #2145.
